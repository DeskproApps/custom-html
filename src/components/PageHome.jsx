import React from 'react';
import PropTypes from 'prop-types';
import Handlebars from 'handlebars';
import { Content } from '@deskpro/apps-components';

/**
 * Renders the defined template.
 */
class PageHome extends React.PureComponent {

  static propTypes = {

    location: PropTypes.object.isRequired,

    history: PropTypes.object.isRequired,

    dpapp: PropTypes.object.isRequired,
  };

  componentDidUpdate()
  {
    const { dpapp } = this.props;
    const { title } = this.props.location.state;
    if (title) {
      dpapp.ui.changeTitle(title);
    }
  }

  render() {
    const { history, dpapp } = this.props;
    const { title, html, context } = this.props.location.state;
    if (!html || !title) {
      history.push("settings");
      history.go(1);
      return null;
    }

    try {
      const template = Handlebars.compile(html);
      const renderedTemplate = template(context);
      return (
        <Content>
          <div dangerouslySetInnerHTML={{ __html: renderedTemplate }} />
        </Content>
      );

    } catch (error) {
      if (error instanceof Error) {
        dpapp.ui.showErrorNotification(error)
      } else {
        dpapp.ui.showNotification(error, 'error')
      }
      console.error(error);
      return null;
    }
  }
}

export default PageHome;
