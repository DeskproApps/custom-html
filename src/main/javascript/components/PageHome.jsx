import React from 'react';
import PropTypes from 'prop-types';
import Handlebars from 'handlebars';
import { sdkConnect } from '@deskpro/apps-sdk-react';
import { Container } from '@deskpro/react-components';

/**
 * Renders the defined template.
 */
class PageHome extends React.PureComponent {
  static propTypes = {
    /**
     * Instance of sdk storage.
     * @see https://deskpro.gitbooks.io/deskpro-apps/api/props/storage.html
     */
    storage: PropTypes.object.isRequired,

    /**
     * Populated with the details of the currently opened ticket.
     * @see https://deskpro.gitbooks.io/deskpro-apps/api/props/tabdata.html
     */
    tabData: PropTypes.object.isRequired,

    /**
     * Instance of sdk ui.
     * @see https://deskpro.gitbooks.io/deskpro-apps/api/props/ui.html
     */
    ui: PropTypes.object.isRequired
  };

  /**
   * Constructor
   *
   * @param {*} props
   */
  constructor(props) {
    super(props);
    try {
      if (!props.storage.app.settings || !props.storage.app.settings.template) {
        props.route.to('settings');
      } else {
        const template = Handlebars.compile(props.storage.app.settings.template || '');
        this.html      = template(props.tabData);
      }
    } catch (error) {
      props.ui.error(error);
      props.route.to('settings');
    }
  }

  /**
   * @returns {XML}
   */
  render() {
    return (
      <Container>
        <div dangerouslySetInnerHTML={{ __html: this.html }} />
      </Container>
    );
  }
}

export default sdkConnect(PageHome);
