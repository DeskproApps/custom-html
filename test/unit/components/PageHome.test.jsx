import React from 'react';
import renderer from 'react-test-renderer';
import { createAppFromProps } from '@deskpro/apps-sdk';
import { createMemoryHistory as createHistory } from "history";


import PageHome from '../../../src/components/PageHome';

it('PageHome renders correctly in the normal state', () => {

  const dpapp = createAppFromProps({
    contextProps : {
      type:       'ticket',
      entityId:   '1',
      locationId: 'ticket-sidebar',
      tabId:      'tab-id',
      tabUrl:     'http://127.0.0.1'
    },
    instanceProps : {
      appId:          '1',
      instanceId:     '1',
      appTitle:       'GitHub',
      appPackageName: 'app-boilerplate-react'
    }
  });

  const history = createHistory({ initialEntries: ["loading"], initialIndex: 0});

  const state = {
    html:   "  Ticket by {{tab.person.emails.[0]}}\n",
    title:  "some title",
    context: {
      tab: {
        person: {
          emails: ["joe.kool@deskpro.com"]
        }
      } ,
      me: {}
    }
  };
  history.push("home", state);
  history.go(1);

  const tree = renderer.create(<PageHome
    dpapp=      {dpapp}
    history=    {history}
    location=   {{ state }}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
