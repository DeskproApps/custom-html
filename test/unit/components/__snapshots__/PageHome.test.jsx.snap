// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHome renders correctly in the normal state 1`] = `
<div
  className="dp-Content"
>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "  Ticket by joe.kool@deskpro.com
      ",
      }
    }
  />
</div>
`;
